File: OOPRecords.Model/Student.cs
using NakedObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace OOPRecords.Model
{
    public class Student
    {
        public TeacherRepository TeacherRepository { set; protected get; }

        public SubjectRepository SubjectRepository { set; protected get; }

        public IDomainObjectContainer Container { set; protected get; }


        [Hidden(WhenTo.Always)]
        public virtual int Id { get; set; }

        [MemberOrder(1)]
        public virtual string StudentNumber { get; set; }

        [MemberOrder(2)]
        public virtual string FirstName { get; set; }

        [MemberOrder(3)]
        public virtual string LastName { get; set; }

        [MemberOrder(4)][Hidden(WhenTo.OncePersisted)]
        public virtual DateTime DateOfBirth { get; set; }

        public void ConfirmDateOfBirth(DateTime dateOfBirth)
        {
            var message = dateOfBirth == DateOfBirth ? "CORRECT" : "INCORRECT";
            Container.InformUser($"The date of birth you entered is {message} for this student.");
        }

        public string ValidateDateOfBirth(DateTime dob)
        {
            return dob > DateTime.Today ? "Date of Birth cannot be after today" : null;
        }


        [MemberOrder(5)]
        public virtual Teacher Tutor { get; set; }

        public IQueryable<Teacher> AutoCompleteTutor([MinLength(3)] string match)
        {
            return TeacherRepository.FindTeacherByLastName(match);
        }


        [MemberOrder(6)]
        [Eagerly(EagerlyAttribute.Do.Rendering)]
        [TableView(false, "Subject", "SetName", "Teacher")]
        public virtual ICollection<Set> Sets { get; set; } = new List<Set>();

        [Hidden(WhenTo.Always)]
        public int Age()
        {
            var today = DateTime.Today;
            int age = today.Year - DateOfBirth.Year;

            if (today.Month < DateOfBirth.Month || (today.Month == DateOfBirth.Month && today.Day < DateOfBirth.Day))
                age--;
            return age;
        }

        public override string ToString()
        {
            return $"{FirstName} {LastName}, Age {Age()}";
        }

        public IQueryable<SubjectReport> ListRecentReports()
        {
            int id = this.Id;
            var studentReps = Container.Instances<SubjectReport>().Where(sr => sr.Student.Id == id);
            return studentReps.OrderByDescending(sr => sr.Date);
        }

        public SubjectReport CreateNewReport()
        {
            var rep = Container.NewTransientInstance<SubjectReport>();
            rep.Student = this;
            return rep;
        }
    }
}

File: OOPRecords.ModelVB/Student.vb
Imports System.ComponentModel.DataAnnotations
Imports NakedObjects

Public Class Student

    Private _teacherRep As TeacherRepository
    Public WriteOnly Property TeacherRepository() As TeacherRepository
        Set(ByVal value As TeacherRepository)
            _teacherRep = value
        End Set
    End Property

    Private _container As IDomainObjectContainer
    Public WriteOnly Property Container() As IDomainObjectContainer
        Set(ByVal value As IDomainObjectContainer)
            _container = value
        End Set
    End Property

    <Hidden(WhenTo.Always)>
    Public Overridable Property Id As Integer

    <MemberOrder(2)>
    Public Overridable Property FirstName As String

    <MemberOrder(3)>
    Public Overridable Property LastName As String

    <MemberOrder(4)> <Hidden(WhenTo.OncePersisted)>
    Public Overridable Property DateOfBirth As Date

    Public Function ValidateDateOfBirth(ByVal dob As DateTime) As String
        Return If(dob > DateTime.Today, "Date of Birth cannot be after today", Nothing)
    End Function

    Public Sub ConfirmDateOfBirth(ByVal dateOfBirth As DateTime)
        Dim message = If(dateOfBirth = dateOfBirth, "CORRECT", "INCORRECT")
        _container.InformUser($"The date of birth you entered is {message} for this student.")
    End Sub

    <MemberOrder(5)>
    Public Overridable Property Tutor As Teacher

    Public Function AutoCompleteTutor(<MinLength(3)> ByVal match As String) As IQueryable(Of Teacher)
        Return _teacherRep.FindTeacherByLastName(match)
    End Function

    <MemberOrder(1)>
    Public Overridable Property StudentNumber As String

    Private _sets As ICollection(Of TeachingSet) = New List(Of TeachingSet)

    <MemberOrder(6)>
    Public Overridable Property Sets() As ICollection(Of TeachingSet)
        Get
            Return _sets
        End Get
        Set(value As ICollection(Of TeachingSet))
            _sets = value
        End Set
    End Property

    <Hidden(WhenTo.Always)>
    Public Function Age() As Integer
        Dim today = Date.Today
        Dim ageToday As Integer = today.Year - DateOfBirth.Year
        If today.Month < DateOfBirth.Month OrElse (today.Month = DateOfBirth.Month AndAlso today.Day < DateOfBirth.Day) Then ageToday -= 1
        Return ageToday
    End Function

    Public Overrides Function ToString() As String
        Dim x = _teacherRep.AllTeachers()
        Return $"{FirstName} {LastName}, Age {Age()}"
    End Function
End Class
